erts/emulator/beam/erl_db.c

```diff
@@ -783,57 +807,35 @@ DbTable* db_get_table_aux(Process *p,
 
     ASSERT((what == DB_READ_TBL_STRUCT) == (kind == LCK_NOLOCK_ACCESS));
 
-    if (META_DB_LOCK_FREE())
-        name_already_locked = 1;
-
     if (is_not_atom(id)) {
         tb = tid2tab(id, &p->fvalue);
     } else {
-	struct meta_name_tab_entry* bucket = meta_name_tab_bucket(id,&name_lck);
-	if (!name_already_locked)
-	    erts_rwmtx_rlock(name_lck);
-	else {
-	    ERTS_LC_ASSERT(META_DB_LOCK_FREE()
-                           || erts_lc_rwmtx_is_rlocked(name_lck)
-                           || erts_lc_rwmtx_is_rwlocked(name_lck));
-            name_lck = NULL;
-	}
+        struct meta_name_tab_bucket* bucket = meta_name_tab_bucket(id);
+        struct meta_name_tab_entries *entries =
+            (struct meta_name_tab_entries*) erts_atomic_read_ddrb(&bucket->entries);
+        Uint bucket_size = erts_atomic32_read_rb(&entries->size);
+
         tb = NULL;
-	if (bucket->pu.tb != NULL) {
-	    if (is_atom(bucket->u.name_atom)) { /* single */
-		if (bucket->u.name_atom == id)
-		    tb = bucket->pu.tb;
-	    }
-	    else { /* multi */
-		Uint cnt = unsigned_val(bucket->u.mcnt);
-		Uint i;
-		for (i=0; i<cnt; i++) {
-		    if (bucket->pu.mvec[i].u.name_atom == id) {
-			tb = bucket->pu.mvec[i].pu.tb;
-			break;
-		    }
-		}
-	    }
-	}
 
-	if (tb == NULL) {
-            if (name_lck)
-                erts_rwmtx_runlock(name_lck);
+        for (Uint i = 0; i < bucket_size; ++i) {
+            if (entries->data[i].name_atom == id) {
+                tb = entries->data[i].tb;
+                break;
+            }
+        }
+
+        if (tb == NULL) {
             p->fvalue = EXI_ID;
-	}
+        }
     }
 
     if (tb) {
         if (what == DB_READ_TBL_STRUCT) {
-            if (name_lck)
-                erts_rwmtx_runlock(name_lck);
             return tb;
         }
 
         DB_HASH_ADAPT_NUMBER_OF_LOCKS(tb);
-	db_lock(tb, kind);
-        if (name_lck)
-            erts_rwmtx_runlock(name_lck);
+        db_lock(tb, kind);
```

Comment:
I've tried getting this to work and I'm hitting a small issue.
The lock is not always taken in db_get_table_aux(): if `what == DB_READ_TBL_STRUCT`, then there is an early return before the lock is taken. The double-check might still work, but I'm struggling to figure out how to make it safe, and ensure that the table does not change under us. Apart from that I've got a patch with the relevant changes (removing the hack for size, and doing an atomic write of 0 to the inline_entry size when removing the only table in it (rather than allocating an empty outlined entry).

Reply:
You should double check the name **after** locking the table. Otherwise the table op may be seen as happening after the table was renamed.

```
ets:rename(aaa, bbb),
ets:insert(bbb, {key, bbb})
```
A concurrent process doing `ets:lookup(aaa, key)` should never get `{key,bbb}` if it never existed in the tabled when it was named `aaa.`

For `DB_READ_TBL_STRUCT` it's enough to do the check without lock. Just to make sure the table got the correct name _now._ `DB_READ_TBL_STRUCT` is only used by yielding `ets:insert` as a first cheap lookup of the table. It will lock the table later and then also double check the name. See comment in function `ets_insert_2_list_lock_tbl`.

Reply:
Thanks for the explanation, I added it to the comment (and moved the double-check to after taking the lock).